### ISSUE TITLE
Deep linking and synthesizing the back stack

### ISSUE THREAD

**Issue by GrahamRogers** [bluelinelabs/Conductor#52]
I want to be able to deep-link into a part of my app. As an example, an email app will want to go directly to an email from a notification. According to https://developer.android.com/training/implementing-navigation/temporal.html#SynthesizeBackStack ([cached version since the docs seem to be down](http://webcache.googleusercontent.com/search?q=cache:https://developer.android.com/training/implementing-navigation/temporal.html#SynthesizeBackStack)), the correct behavior is to add to the back stack so that it's as though the user navigated there manually. Essentially, I want to be able to add multiple Controllers to the back stack in one transaction. I could do a bunch of pushController calls in one go, but that seems wasteful since (if I understand the implementation correctly) each controller will have to create its View.

[Flow](https://github.com/square/flow) solves this by using "key" objects to reference individual parts of the back stack. Basically immutable objects. It's then possible to create a History object, which is essentially a list of keys, and use that as the new backstack. Could we have something similar, maybe a `Router.setBackStack(List<Controller>, ControllerChangeHandler)` method?

**Comment by EricKuck** [comment]
This is a great idea. I think the call should be `setBackstack(List<RouterTransaction>)` so we can get nice things like tags and pop handlers built in there too. Any thoughts on what should happen if the developer calls this while there are already controllers on the backstack? Should it just immediately remove all those and replace it with the new backstack? Is there ever a situation where you'd want to transition between the old top controller and the new one?

**Comment by leonardo2204** [comment]
That would be great.
I think there should be 2 methods, `setBackstack(List<RouterTransaction>)` and `setBackstack(List<RouterTransaction>, boolean overrideCurrentBackstack)`, where the default behaviour should not override the Backstack. That's my opinion.
Another point I think it's worth mentioning is, after setting the backstack, the first (actually last, since it's a stack) Controller gets pushed or not ? Or the RootController will be pushed ?

**Comment by hannesstruss** [comment]
I think most intuitive would be to always replace the backstack when calling `setBackstack`, but to allow obtaining a copy of the current backstack and using that as a base to build upon (See #56).

**Comment by hannesstruss** [comment]
A use case where a transition between old and new state would be useful:

```
                      [Wizard Screen 1] -- [Wizard Screen 2]
                     /                                      \
...[Screen A] -- [Screen B]                                 [Screen B']
```

A multi-screen wizard is started from `Screen B`. When the wizard is completed, the user should land on `Screen B'` which is `Screen B` with a new set of arguments. Hitting back on `Screen B'` should navigate back to `Screen A`.

Could be implemented by obtaining the backstack, removing the wizard screens and `Screen B`, adding `Screen B'` and calling `setBackstack` with the new stack, in which case a transition should play.

This could also be achieved by adding a more high level property for `RouterTransaction` (equivalent to `android:launchMode="singleTop"`), but more fine-grained control would be great.

**Comment by EricKuck** [comment]
Good stuff @hannesstruss. I've implemented #56 locally and am working on transitioning between different sets of backstacks.

**Comment by EricKuck** [comment]
This is now available in the v2 release candidate. Updated docs can be found at https://github.com/bluelinelabs/Conductor/blob/v2/README.md. Feel free to suggest any changes before before we officially go to v2!
